localstack/services/stepfunctions/asl/component/intrinsic/function/statesfunction/generic/string_format.py

```diff
@@ -0,0 +1,72 @@
+import json
+from typing import Any, Final
+
+from localstack.services.stepfunctions.asl.component.intrinsic.argument.function_argument_list import (
+    FunctionArgumentList,
+)
+from localstack.services.stepfunctions.asl.component.intrinsic.argument.function_argument_string import (
+    FunctionArgumentString,
+)
+from localstack.services.stepfunctions.asl.component.intrinsic.function.statesfunction.states_function import (
+    StatesFunction,
+)
+from localstack.services.stepfunctions.asl.component.intrinsic.functionname.state_function_name_types import (
+    StatesFunctionNameType,
+)
+from localstack.services.stepfunctions.asl.component.intrinsic.functionname.states_function_name import (
+    StatesFunctionName,
+)
+from localstack.services.stepfunctions.asl.eval.environment import Environment
+
+
+class StringFormat(StatesFunction):
+    _DELIMITER: Final[str] = "{}"
+
+    def __init__(self, arg_list: FunctionArgumentList):
+        super().__init__(
+            states_name=StatesFunctionName(function_type=StatesFunctionNameType.Format),
+            arg_list=arg_list,
+        )
+        if arg_list.size == 0:
+            raise ValueError(
+                f"Expected at least 1 argument for function type '{type(self)}', but got: '{arg_list}'."
+            )
+        if not isinstance(arg_list.arg_list[0], FunctionArgumentString):
+            raise ValueError(
+                f"Expected the first argument for function type '{type(self)}' to be a string, but got: '{arg_list.arg_list[0]}'."
+            )
+
+    def _eval_body(self, env: Environment) -> None:
+        # TODO: investigate behaviour for incorrect number of arguments in string format.
+        self.arg_list.eval(env=env)
+
+        values: list[Any] = list()
+        for _ in range(self.arg_list.size):
+            values.append(env.stack.pop())
+        string_format: str = values.pop()
+        values.reverse()
+
+        string_format_parts: list[str] = string_format.split(self._DELIMITER)
+        string_result: str = ""
+        for part in string_format_parts:
+            string_result += part
+            if values:
+                value = values.pop()
+                if not isinstance(value, str):
+                    value = self._json_to_literal(value)
+                string_result += value
```

Comment:
Could be refactored a bit to be more idiomatic IMO.

A suggestions/nits:
- check the length of values & placeholders explicitly.
- first fix all the values (_json_to_literal) in the same step as you're reversing it, and then either doing the string-formatting loop it in a simpler imperative style with repeated `replace` calls (e.g. `string_format = string_format.replace("{}", value, 1)` or in a maybe more pythonic way via list expressions where you get a list of parts without the placeholders (your `string_format_parts`) and the values, zip  them and then just join them into a string again.